executor/calibrate_resource.go

```diff
@@ -99,7 +166,94 @@ func (e *calibrateResourceExec) Next(ctx context.Context, req *chunk.Chunk) erro
 
 	exec := e.ctx.(sqlexec.RestrictedSQLExecutor)
 	ctx = kv.WithInternalSourceType(ctx, kv.InternalTxnOthers)
+	if len(e.optionList) > 0 {
+		return e.dynamicCalibrate(ctx, req, exec)
+	}
+	return e.staticCalibrate(ctx, req, exec)
+}
 
+func (e *calibrateResourceExec) dynamicCalibrate(ctx context.Context, req *chunk.Chunk, exec sqlexec.RestrictedSQLExecutor) error {
+	startTs, endTs, err := e.checkDynamicCalibrateOptions()
+	startTime := startTs.In(e.ctx.GetSessionVars().Location()).Format("2006-01-02 15:04:05")
+	endTime := endTs.In(e.ctx.GetSessionVars().Location()).Format("2006-01-02 15:04:05")
+	if err != nil {
+		return err
+	}
+	totalKVCPUQuota, err := getTiKVTotalCPUQuota(ctx, exec)
+	if err != nil {
+		return err
+	}
+	totalTiDBCPU, err := getTiDBTotalCPUQuota(ctx, exec)
+	if err != nil {
+		return err
+	}
+	rus, err := getRUPerSec(ctx, exec, startTime, endTime)
+	if err != nil {
+		return err
+	}
+	tikvCPUs, err := getComponentCPUUsagePerSec(ctx, exec, "tikv", startTime, endTime)
+	if err != nil {
+		return err
+	}
+	tidbCPUs, err := getComponentCPUUsagePerSec(ctx, exec, "tidb", startTime, endTime)
+	if err != nil {
+		return err
+	}
+	quotas := make([]float64, 0)
+	lowCount := 0
+	tidbCPULowCount := 0
+	tikvCPULowCOunt := 0
```

Comment:
maybe can  reduce the initial number of rows
```suggestion
	lowCount, tidbCPULowCount, tikvCPULowCount := 0, 0, 0
```
BTW, `tikvCPULowCOunt` has a extra big `O`